extmod/modselect.c

```diff
@@ -107,6 +268,69 @@ STATIC mp_uint_t poll_map_poll(mp_map_t *poll_map, size_t *rwx_num) {
     return n_ready;
 }
 
+STATIC mp_uint_t poll_group_poll_all(poll_group_t *poll_group, size_t *rwx_num, mp_uint_t timeout) {
+    mp_uint_t start_ticks = mp_hal_ticks_ms();
+
+    #if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+
+    for (;;) {
+        MP_THREAD_GIL_EXIT();
+
+        // Compute the timeout.
+        int t = 1;
+        if (poll_group_all_are_fds(poll_group)) {
+            if (timeout == (mp_uint_t)-1) {
+                t = -1;
+            } else {
+                mp_uint_t delta = mp_hal_ticks_ms() - start_ticks;
+                if (delta >= timeout) {
+                    t = 0;
+                } else {
+                    t = timeout - delta;
+                }
+            }
+        }
+
+        // Call system poll for those objects that have a file descriptor.
+        int n_ready = poll(poll_group->pollfds, poll_group->len, t);
+
+        MP_THREAD_GIL_ENTER();
+
+        if (n_ready == -1) {
+            int err = errno;
+            if (err != EINTR) {
```

Comment:
should this also be `EAGAIN` ?

Also, the old code used `MP_HAL_RETRY_SYSCALL` ?

Reply:
Comment added about PEP 475 and `MP_HAL_RETRY_SYSCALL`.